crates/matrix-sdk/src/test_utils/mocks.rs

```diff
@@ -0,0 +1,389 @@
+// Copyright 2024 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![allow(missing_debug_implementations)]
+
+use std::sync::{Arc, Mutex};
+
+use matrix_sdk_base::deserialized_responses::TimelineEvent;
+use matrix_sdk_test::{
+    test_json, InvitedRoomBuilder, JoinedRoomBuilder, KnockedRoomBuilder, LeftRoomBuilder,
+    SyncResponseBuilder,
+};
+use ruma::{OwnedEventId, OwnedRoomId, RoomId};
+use serde_json::json;
+use wiremock::{
+    matchers::{header, method, path, path_regex},
+    Mock, MockBuilder, MockGuard, MockServer, Respond, ResponseTemplate, Times,
+};
+
+use super::logged_in_client;
+use crate::{Client, Room};
+
+pub struct MatrixMockServer {
+    server: MockServer,
+    client: Client,
+
+    /// Make the sync response builder stateful, to keep in memory the batch
+    /// token and avoid the client ignoring subsequent responses after the first
+    /// one.
+    sync_response_builder: Arc<Mutex<SyncResponseBuilder>>,
+}
+
+impl MatrixMockServer {
+    /// Create a new mocked server specialized for Matrix usage.
+    pub async fn new() -> Self {
+        let server = MockServer::start().await;
+        let client = logged_in_client(Some(server.uri().to_string())).await;
+        Self { client, server, sync_response_builder: Default::default() }
+    }
+
+    /// Creates a new [`MatrixMockServer`] when both parts have been already
+    /// created.
+    pub fn from_pair(server: MockServer, client: Client) -> Self {
+        Self { client, server, sync_response_builder: Default::default() }
+    }
+
+    /// Return the underlying client.
+    pub fn client(&self) -> Client {
+        self.client.clone()
+    }
+
+    /// Return the underlying server.
+    pub fn server(&self) -> &MockServer {
+        &self.server
+    }
+
+    /// Overrides the sync/ endpoint with knowledge that the given
+    /// invited/joined/knocked/left room exists, runs a sync and returns the
+    /// given room.
+    pub async fn sync_room(&self, room_id: &RoomId, room_data: impl Into<AnyRoomBuilder>) -> Room {
```

Comment:
Trying to think of a more informative name for this...

Maybe `create_or_change_room` ? yuck.

Reply:
`mock_and_run_sync_for_room`? :woozy_face: